More parameterizable extract-lines

kba opened the pull request:
* Make the thresholds for line length, width and height configurable
* Get the presentation URL for the page for both SLUB and SBB works

wrznr committed:
initial version

Commit:
use JSON files instead of PAGE-XML:

- CLI: glob *.json instead of page/*.xml, then load JSON
  and pass to Dataset.add_image as annotation
- CLI: fix argument type constraints (dir/file/exists), add help
- Dataset: instead of parsing PAGE-XML via lxml, use metadata
  in JSON annotation to get region coordinates and region types;
  use image file to get image dimensions; keep generating bboxes
  from polygons and only text class (but prepare for more)
- make pylint less unhappy

Commit:
moved files to directory maskrcnn-cli

Commit:
Merge branch 'maskrcnn-cli'

Commit:
fix bbox_from_polygon initvals

Commit:
fg vs bg instead of text vs other

Commit:
evaluate: show actual filenames, filter out NaN results

Commit:
train: avoid loading head layers only for transfer from COCO

Commit:
train: use early stopping and store best weights to file

Commit:
use polygon as mask instead of full bbox

Commit:
predict: show (transparent) pixel mask, show filename, save PNG files

Commit:
enable all classes

Commit:
predict: no title/legend, no border around image

Commit:
train/evaluate: avoid samples without regions

Commit:
Merge remote-tracking branch 'origin/master' into maskrcnn-with-classes

Commit:
no downscaling during prediction

Commit:
training: increase EarlyStopping patience

Commit:
training: expose layer depth parameter (not just heads vs all)

Commit:
add mask-rcnn dependency

Commit:
script to train PubLayNet directly from MS-COCO json

Commit:
Merge remote-tracking branch 'upstream/master' into maskrcnn-with-classes

bertsky committed:
Merge pull request #34 from OCR-D/maskrcnn-with-classes

Mask-RCNN with classes and more CLI options

Commit:
improve PubLayNet COCO script...

- further reduce required GPU memory
  in MaskRCNN configuration
- further adapt from COCO to PubLayNet:
  - remove all remaining COCO idiosyncracies
  - automatic download (and progress bar)
  - separate JSON path from parent directory
    path for images for full flexibility
  - use sub-parser for CLI options diverging
    between commands, improve descriptions
- add command `test` for PubLayNet (reading
  val.json for classes and test/*.jpg for
  prediction, writing test.json for results)
- add command `predict` for arbitrary files
  (reading val.json for classes and filenames
  for prediction, writing *.pred.png files
  with masks plotted in pseudo-color)
- more converter routines:
  - `showAnns`: (modified) version of pycocotools
     to plot predictions (PNG of original size
     and in pseudo-color)
  - `test_coco`: apply on Dataset
  - `load_files`: Dataset from arbitrary files
  - `dump_coco`: Dataset to COCO
  - COCO JSON serialization

Commit:
Merge branch 'master' into maskrcnn-cli

Commit:
new partial (re)segmentation model for addresses...

- ocrd-segment-extract-address: export mask+JSON data
  from GT with text region classification for addresses,
  represent text classes in alpha channel of images
- address: train/predict/evaluate visual address segment
  detection model based on alpha-marked input, in COCO
  format (similar to PubLayNet);
  when transferring weights from COCO model initially,
  exclude conv1 layer and train it from scratch along
  with MRCNN heads
- ocrd-segment-detect-address: load address detection
  model, process full segmented and recognized input,
  calling text classifier on possible address parts
  and providing alpha-masked page images to the model,
  run detection and post-process each region candidate,
  keeping only 1-best for rcpt/sndr (but many for contact),
  and dilate mask until single contour can be found;
  remove overlapping existing regions, taking their text
  lines and region order position

Commit:
detect-address: use 3-class model

Commit:
detect-address: minor fixes...

- avoid predictions that overlap each other
  (use the one with best score)
- try harder not to generate invalid polygons
  (both from raw detection and from disjunction
   with overlapping text line)
- use larger extent / IoU around region candidates
  (to check if they can replace existing regions)
- avoid removing the same text region twice
  (because it contains multiple address regions)
- rename stolen text lines
- avoid calling text classifier on too short text

Commit:
detect_address: rely more on text line classification results...

- annotate results of text line classification via @custom
- avoid region candidates that do not overlap any text lines
  with address classes (ghosts due to unrealistic training)

Commit:
detect_address: use proper reading-order iterator...

(when traversing the text lines of the input, sort them
 according to their ReadingOrder position if available,
 to ensure text classification gets to see address parts
 in the right order)

Commit:
split detect_address into classify_address_{text,layout}

Commit:
address model: use 2 channels instead of alpha directly...

(after importing images, split alpha channel into
 text/non-text and address/non-address to make
 models more robust)

Commit:
Merge branch 'master' into maskrcnn-cli

Commit:
Merge branch 'master' into maskrcnn-cli

Commit:
classify_address_text: add some text normalization heuristics

Commit:
classify_address_layout: fix 9f99bfc…

- avoid removing the same text region twice
 (because it contains multiple address regions)

Commit:
classify_address_*: improve inference…

- use new generalized region iterators from core
- for text classification,
  - be a little robust against OCR errors
  - normalize various kinds of field separators
  - allow longer addresses when concatenated
  - concatenate up to 3 previous lines instead of 2
  - avoid adding field separator iff
    - it is already a full address or
    - previous line is left of current line

Commit:
address: use augmentation, allow empty pages

(employ augmenters with segmentation maps to
 make network robust against some classes not
 always appearing both in context and target,
 against some lines not appearing in context,
 or against blacked out input)

Commit:
address: simplify CLI and implementation…

- unify test/predict command: read categories from existing JSON,
  filling images and annotations; do not plot by default anymore
- add switch for plotting files (both test and evaluate),
  allow setting filename suffix
- fix for COCO export: ensure successive annotation IDs
- allow setting active class IDs during prediction
- evaluate and test share same basic loop

Commit:
coco-categories: tool for spawning empty dataset with same categories

Commit:
OCR-D address CLIs: update, more robust…

- rebase on new core utils for METS ID generation etc
- rebase on new core recursive region enumerator
- ensure all input image types work (Pillow putalpha)
- pass in GT categories to extract as extra parameter

Commit:
replace_page: new OCR-D CLI (inverse of replace_original)

Commit:
formdata: variant of address segmentation…

- same basic partial segmentation scheme
  (marking context as extra channel in the input),
  but targeted at arbitrary forms/tables instead of
  address regions, and segmenting one class at a time
  (marking its specific context) anywhere on the page
  with up to 1 instance instead of multi-class
  (on uniform context) doing re-segmentation only but
  with any number of instances
- train and predict one class at a time (suppressing
  other outputs), but on shared model
- many more classes, defined independent of occurance
  in COCO sources; active class determined from COCO
  source (both at training and inference, i.e. empty
  source is needed for new files)
- GT uses arbitrary PAGE region types (but no @custom types)
- different augmentation (obviously)

Commit:
Merge remote-tracking branch 'upstream/master' into maskrcnn-cli

Commit:
replace_page: catch invalid coords and skip page

Commit:
extract_{address,formdata}: if text region has no text lines, make a fill-dummy

Commit:
{address,formdata}: add --epochs option, save last weights file under dataset basename

kba committed:
getLogger per method

bertsky committed:
Update CHANGELOG.md

bertsky committed:
Update CHANGELOG.md

bertsky committed:
:package: 0.1.3

bertsky committed:
Update CHANGELOG.md

Commit:
{address,formdata}: fix showAnns/plot_result (static functions)

Commit:
formdata: allow joint training/evaluation on multiple COCO datasets

Commit:
address: allow joint training/evaluation on multiple COCO datasets

Commit:
formdata: detect up to 1 instance only

Commit:
extract_address: allow mapping normal TextRegion types besides custom

Commit:
getLogger per method

Commit:
install imgaug from source instead of wheel…

…(to avoid dragging in opencv-python when we
actually depend on opencv-python-headless already,
which silently clashes)

Commit:
hide imgaug requirement (relying on mask-rcnn instead)

Commit:
repair: fix coordinate consistency/validity errors if possible

Commit:
replace_page: delegate to repair.ensure_consistent

Commit:
replace_page: delegate to zip_input_files now in core

Commit:
replace_original: delegate to repair.ensure_consistent

Commit:
evaluate: delegate to zip_input_files now in core

Commit:
Merge remote-tracking branch 'origin/excel_gt_writer' into maskrcnn-cli

Commit:
repair: also fix negative coords, also on page level

Commit:
Merge branch 'maskrcnn-cli' of https://github.com/OCR-D/ocrd_segment into maskrcnn-cli

Commit:
classify_address_layout: fix marking non-address text

Commit:
add entry points for stand-alone CLIs

Commit:
formdata/address: allow setting split ratio and seed value

Commit:
extract_address: write COCO JSON to workspace directory

Commit:
address/formdata: fix help text formatting (escape %) etc

Commit:
classify_address_layout: improve post-processing/decoding…

- ensure valid and consistent (with parent) polygons
- avoid assigning neighbours too far away from mask
- extend mask by morphological criterion: hull of all
  overlapping connected components in binarized image
- don't suppress non-best instances where overlap is small

Commit:
classify_address_layout: reassign non-best sndr/rcpt instances to contact

Commit:
replace-original: invalidate Border/orientation, add new original as derived image too

Commit:
add 3 new processors `classify-*-{text,layout,dummy}:

- classify-formdata-text: use fuzzy string matching
  against predefined keywords for all classes to
  mark lines/words as context (via @custom), and to
  mark lines/words as target for text-only classes
- classify-formdata-layout: detect target regions
  from context with Mask-RCNN model; for each page,
  collect all context classes among lines/words and
  prepare a corresponding 5ch-image with text+context
  marks; pack into batches, predict, unpack; apply
  post-processing (connected components, NMS, 1-best)
  and annotate result as target (via @custom)
- classify-formdata-dummy: align original segmentation
  with GT annotations for all classes into a single
  PAGE with context/target (via @custom)
- formdata: 2 new classes (plus upcoming classes
  as comments)

Commit:
replace-original: also replace @width/height for clarity

Commit:
classify-address-layout: use binarized (not raw) input images for prediction

Commit:
extract-address: also read marks in TextLines; add default categories mapping

Commit:
replace-original: only ensure coord validity, not consistency

Commit:
repair: also ensure >2 points

Commit:
address: fix path for saving last trained model

Commit:
address/formdata: allow setting different learning rate

Commit:
classify_formdata_text: improve keywords, match numbers and short strings differently

Commit:
extract_lines: ensure output fileGrp exists

Commit:
extract_lines: try to guess presentation URL (Kitodo/SLUB)

Commit:
extract_lines (excel worksheet for GT annotation): improve…

- freeze first (header) row
- put hard to type historical characters in first row
- ensure line images fit into rows:
  - instead of arbitrary image size and 3-row gap,
    scale to fit exactly one row
  - increase row height
- make line images clickable (href to presentation URL)
- do not include lines with too few/short text or
  too small line images

Commit:
Merge remote-tracking branch 'refs/remotes/origin/maskrcnn-cli' into maskrcnn-cli

bertsky committed:
replace-page: allow non-PAGE inputs, too

(fixes regression in 06c52f3feb0f200a2853c97ef5fbccd0edffc38a)

Commit:
repair: update docstring

kba committed:
extract-lines: introduce parameter "library-convention"

kba committed:
extract-lines: make min-line-{length,width,height} parameterizable

kba committed:
implement "sbb" library convention for presentation URL

kba committed:
put symbols into one cell each

bertsky commented on ocrd_segment/extract_lines.py:
```diff
@@ -103,39 +109,20 @@ def process(self):
-            worksheet.write('E1', 'ſ ꝛ aͤ oͤ uͤ æ œ Æ Œ ℳ  ç ę ë - ⸗ = Α α Β β ϐ Γ γ Δ δ Ε ε ϵ Ζ ζ Η η Θ θ ϑ Ι ι Κ κ ϰ Λ λ Μ μ Ν ν Ξ ξ Ο ο Π π ϖ Ρ ρ ϱ Σ σ 	ς ϲ Τ τ Υ υ ϒ Φ φ ϕ Χ χ Ψ ψ Ω ω')
+            symbols = 'ſ ꝛ aͤ oͤ uͤ æ œ Æ Œ ℳ  ç ę ë - ⸗ = Α α Β β ϐ Γ γ Δ δ Ε ε ϵ Ζ ζ Η η Θ θ ϑ Ι ι ' \
+                      'Κ κ ϰ Λ λ Μ μ Ν ν Ξ ξ Ο ο Π π ϖ Ρ ρ ϱ Σ σ ς ϲ Τ τ Υ υ ϒ Φ φ ϕ Χ χ Ψ ψ Ω ω'.split(' ')
+            for i, s in enumerate(symbols):
+                col_idx = 3 + i
+                worksheet.write_string(0, col_idx, s)
+                worksheet.set_column(col_idx, col_idx, 2)
```
Good idea! Will the line images still stretch across all these columns if necessary?

kba replied:
Yes, the line image will stretch at least in my tests in Excel 2016 and LibreOffice 5. I had also prepared a solution with macros assigned to buttons so that you just need to click a "virtual key" to copy the glyph into clipboard. But macros are a source of vulnerabilities and it's hard to to argue for enabling them.

Another idea was to have a dead simple HTML+Javascript page with a virtual keyboard of click-to-copy symbols to have open in a split browser window.

But let's first test how this "one-char-per-cell" works, ergonomically.

bertsky replied:
Absolutely!

bertsky commented on ocrd_segment/extract_lines.py:
```diff
@@ -248,3 +237,42 @@ def process(self):
+            return self._get_presentation_image_slub(input_file)
+        elif library_convention == 'sbb':
+            return self._get_presentation_image_sbb(input_file)
+        raise NotImplementedError("Unsupported library convention '%s'" % library_convention)
```
I suggest we allow setting _no_ convention, too. After all, clickable images are just a comfort feature.

kba replied:
True, I'll add that.

kba replied:
b876574 the default is now none, to just add an empty string as the URL. If that breaks your scripts, I can change default back to `slub`.

bertsky replied:
It's a parameter now, so no problem.

I just wonder whether (after merging this) I should rebase that branch on master, stripping out the `maskrcnn-cli` merge. After all, one can always merge branches in a local deployment when both features are needed, but not everyone wants all those extra dependencies (at least for now)...

kba replied:
I had no problem installing the dependencies, that the `maskrcnn` clones the github repository on every install is a bit annoying.

Do you want to do the cherry-picking or shall I?

bertsky replied:
> that the `maskrcnn` clones the github repository on every install is a bit annoying.

Yes. It's because of the `git+https` spec in my requirements – I have to use my own fork, but don't want to publish under PyPI (yet). I'd have to come up with a good alternative package name for that, but since the original is unmaintained now...

I guess I could try adding a fixed git tag to the URL, but IIUC the actual reason for re-installation each time is that I don't use `--editable` ...

> Do you want to do the cherry-picking or shall I?

No thanks, IIRC I had some merge conflicts earlier – this I have to fiddle together by myself.

kba replied:
> I guess I could try adding a fixed git tag to the URL

Don't bother, it will still clone the repo to get that tag in my experience (it's a limitation of git).

> No thanks, IIRC I had some merge conflicts earlier – this I have to fiddle together by myself.

I dodged a bullet there :) Thanks!

bertsky commented on ocrd_segment/ocrd-tool.json:
```diff
@@ -333,6 +333,27 @@
+          "type": "string",
+          "enum": ["slub", "sbb"],
+          "default": "slub",
+          "description": "Libraries have slightly different conventions in their METS files"
```
```suggestion
          "description": "Scheme to reconstruct presentation URL of original pages, used to make line images hyperlinked. Libraries have different conventions in their METS files. Set to none to disable."
```

bertsky reviewed (changes requested):
Very helpful, thanks!

kba committed:
Update ocrd_segment/ocrd-tool.json

Co-authored-by: Robert Sachunsky <38561704+bertsky@users.noreply.github.com>

kba committed:
extract-lines: offer a no-op library convention "none" w/o presentation image URL

bertsky commented on ocrd_segment/extract_lines.py:
```diff
@@ -109,7 +109,7 @@ def process(self):
             worksheet.write('C1', 'Status', bold)
             worksheet.write('D1', 'Image', bold)
-            symbols = 'ſ ꝛ aͤ oͤ uͤ æ œ Æ Œ ℳ  ç ę ë - ⸗ = Α α Β β ϐ Γ γ Δ δ Ε ε ϵ Ζ ζ Η η Θ θ ϑ Ι ι ' \
+            symbols = 'ſ ꝛ 𝔷 aͤ oͤ uͤ æ œ Æ Œ ℳ  ç ę ë - ⸗ = Α α Β β ϐ Γ γ Δ δ Ε ε ϵ Ζ ζ Η η Θ θ ϑ Ι ι ' \
```
I am quite surprised by this. This character registers as `MATHEMATICAL FRAKTUR SMALL Z`. So outside of modern-day formula typesetting, this should not play any role. Of course there is a full set of Unicode codepoints with Fraktur glyphs for mathematical purposes, but my understanding was that historical OCR (and DTA/OCR-D transcription rules in particular) should not use these at all. Fraktur is but a font in most contexts, not a script. (I know some contemporary/computer-era Fraktur enthusiasts nevertheless use these for typesetting universally, but that's just a hack IMHO.) 

Conversely, if `𝔷`, then why not also `𝔐` or `𝕼`?

(I know, you might argue the same with the above `SCRIPT CAPITAL M`, which I found quite frequently in GT4HistOCR – as currency sign for marks. So you could equally say, where's `ℓ`, `ℛ`, `ℱ` etc? Answer: I am happy to include them here if someone can point me to actual use in historic texts.)

kba replied:
Hah, sry, I did **not** mean to push this change, I share your skepticism towards (mis)using mathematical symbols for transcription of Fraktur but included it for discussion with our industrious transcribers. I'll force push it out of this PR.

But what I want to add is to change the default font. At least on our systems, the default is `Calibri` which is not very readable (`I` vs `l` etc.) and doesn't display `ꝛ` (r rotunda) correctly. From the list of built-in fonts, `Segoe UI Semibold` looks the most readable and has a fairly wide support of characters AFAICT.

bertsky replied:
> Hah, sry, I did **not** mean to push this change, I share your skepticism towards (mis)using mathematical symbols for transcription of Fraktur but included it for discussion with our industrious transcribers. I'll force push it out of this PR.

I see. Thanks!

> But what I want to add is to change the default font. At least on our systems, the default is `Calibri` which is not very readable (`I` vs `l` etc.) and doesn't display `ꝛ` (r rotunda) correctly. From the list of built-in fonts, `Segoe UI Semibold` looks the most readable and has a fairly wide support of characters AFAICT.

Oh, sorry – I did not even notice `Calibri` was the default font for spreadsheets. Since I don't have any M$ fonts installed whatsoever, and my fallbacks all look good, I blindly assumed there was no problem in that regard. But judging by font preview on the web, `Segoe UI Semibold` also does not cover all the glyphs we want here (e.g. no `ℳ` or `⸗`).

Could we please try to find some widely available, ideally free font family?

kba replied:
> Oh, sorry – I did not even notice Calibri was the default font for spreadsheets.

It is in Excel. In LibreOffice, the default font `Liberation Sans` displays all characters just fine.

> But judging by font preview on the web

There are fallbacks for those characters not in the font. Don't ask me how M$ font layering works but at least all characters are properly displayed. `I` vs `l` etc. are still a problem though.

> Could we please try to find some widely available, ideally free font family?

I have heard good things about https://www.pgdp.net/wiki/DP_Sans_Mono But the issue is more that we must work with what we have, having users install additional fonts is problematic. If we could bundle the font with the tables, while a waste of space, would be best for interoperability, haven't researched whether that's possible though, I doubt it though :(

bertsky replied:
Understood. So we already have quite some complexity here: typical preinstalled fonts available in various OS, fallback/layering in OS' font servers, and even inter-operability (if someone picks up a file with Windows/Excel and makes edits, then sends it to me for confirmation/repairs, I'll see the `Calibri` fallback instead of the `Liberation Sans` default, and vice versa...)

Yes, I agree we should use some monospace font for our purposes here. But `DP Sans Mono` (like `DejaVu Sans Mono` and even `Liberation Sans Mono` or `Monospace Regular`) does not render superscript umlauts well. Also, I wonder whether sans-serif is really what we want. I found both `Latin Modern Mono` (from LaTeX) and `Ubuntu Mono` quite acceptable.

I agree we should try to work towards maximal compatibility with pre-installed fonts. But I have no idea how OSs configure family fallbacks. 

I think we could "bundle" a ttf or otf file along with the generated xlsx files for users who are willing to put up with right-clicking and installing (should work on all major platforms by now), though.

kba replied:
I wish it were that easy but Windows is a terrible OS as we all know and agree on ;) This is what happens when I try "right-clicking and installing":

![image](https://user-images.githubusercontent.com/273367/105730705-efb9d600-5f2e-11eb-93a5-77b0c860008d.png)

And if I fail to install a font, I doubt my colleagues will succeed. I remember having to resort to a local Administrator account and hours of trial and error to get a functional monospace font installed.

bertsky replied:
> I wish it were that easy but Windows is a terrible OS as we all know and agree on ;) This is what happens when I try "right-clicking and installing":
> 
[...]
> 
> And if I fail to install a font, I doubt my colleagues will succeed. I remember having to resort to a local Administrator account and hours of trial and error to get a functional monospace font installed.

Wow, mind-boggling. I had no idea Windows is **that** much of a failure to this day.

Commit:
extract_lines (xlsx writer): no formula/nr interpretation

Commit:
extract_lines (xlsx writer): protect sheet (only text/status editable)

Commit:
extract_regions: fix to avoid zero-size images

bertsky commented:
Sorry, did not find yet the time to rework the `excel_gt_writer` to exclude the commits from `maskrcnn-cli` branch again.

But there was another flaw in my version: text lines starting with `=` were interpreted as formulas. Also, I've decided to protect all parts of the sheet not meant for editing. So you might want to cherry-pick from upstream for the time being...

kba committed:
Merge remote-tracking branch 'origin/excel_gt_writer' into excel-cntd

# Conflicts:
#	ocrd_segment/extract_lines.py

bertsky commented:
Note: I have rebased the `excel_gt_writer` branch such that `maskrcnn-cli` is kept out, then applied your changes by cherry-picking and resolved conflicts. I have further introduced a parameter `output-types`.

Do you think we could merge this into `master` permanently?

kba commented:
> Note: I have rebased the `excel_gt_writer` branch such that `maskrcnn-cli` is kept out, then applied your changes by cherry-picking and resolved conflicts. I have further introduced a parameter `output-types`.
> 
> Do you think we could merge this into `master` permanently?

We moved to using https://github.com/qurator-spk/neat for our transcriptions because as a web-based tool, the font problems were easy to fix, it has a neater (pun intended) UI than Excel and is a single file to be opened in any browser.

So while I think it's useful functionality with few added dependencies and is a workable solution for low-tech transcriptions, it is not essential for us. But it might be to others, so +1 for merging into main/master.
